lister.cc

```diff
@@ -75,3 +77,44 @@ future<> lister::rmdir(fs::path dir) {
         return remove_file(dir.native());
     });
 }
+
+future<std::optional<directory_entry>> queue_lister::get() {
+    if (!_done) {
+        auto f = co_await open_checked_directory(general_disk_error_handler, _dir.native());
+        auto walker = [this] (fs::path dir, directory_entry de) {
+            return _q.push_eventually(std::make_optional<directory_entry>(std::move(de))).handle_exception([] (std::exception_ptr) {
+                // ignore exception, queue was aborted
+            });
+        };
+        assert(!_lister);
+        _lister = std::make_unique<lister>(std::move(f), _type, std::move(walker), _filter, _dir, _do_show_hidden);
+        auto done_fut = _lister->done().then_wrapped([this] (future<> f) {
+            if (!f.failed()) {
+                return _q.push_eventually(std::nullopt).handle_exception([] (std::exception_ptr) {
+                    // ignore exception, queue was aborted
+                });
+            }
+            _q.abort(f.get_exception());
+            return make_ready_future<>();
+        }).finally([this] {
+            _lister.reset();
+        });
+        _done = std::make_optional<future<>>(std::move(done_fut));
+    }
+    co_return co_await _q.pop_eventually();
```

Comment:
Unless I missed something you're not using coroutines here for anything else, so you can do more simply (and I think more efficiently?)
`return _q.pop_eventaully();`

Reply:
I call `co_await open_checked_directory` above...

Reply:
So I did miss something :-)